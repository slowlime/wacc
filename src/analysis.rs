mod class_resolver;
pub mod error;
mod typeck;
mod typectx;
mod validate;

pub use error::TypeckError;
pub use typeck::{TypeChecker, TypeckResult};
pub use typectx::TypeCtx;
pub use validate::validate_classes;
