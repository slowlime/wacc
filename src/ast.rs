use crate::token::{Symbol, Token};

macro_rules! define_op_kind {
    ($name:ident { $( $op:ident => $symbol:ident, )+ }) => {
        #[derive(Debug, Clone, Copy, Eq, PartialEq, Hash)]
        pub enum $name {
            $( $op, )+
        }

        impl $name {
            pub fn symbol(self) -> Symbol {
                match self {
                    $( Self::$op => Symbol::$symbol, )+
                }
            }
        }

        impl TryFrom<Symbol> for $name {
            type Error = ();

            fn try_from(symbol: Symbol) -> Result<Self, Self::Error> {
                match symbol {
                    $( Symbol::$symbol => Ok(Self::$op), )+
                    _ => Err(()),
                }
            }
        }
    }
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct Program<'a> {
    pub classes: Vec<Class<'a>>,
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct Class<'a> {
    pub name: Name<'a>,
    pub inherits: Option<Name<'a>>,
    pub features: Vec<Feature<'a>>,
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct Name<'a>(pub Token<'a>);

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub enum Feature<'a> {
    Method(Method<'a>),
    Field(Binding<'a>),
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct Method<'a> {
    pub name: Name<'a>,
    pub params: Vec<Formal<'a>>,
    pub return_ty: Name<'a>,
    pub body: Expr<'a>,
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct Formal<'a> {
    pub name: Name<'a>,
    pub ty: Name<'a>,
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct Binding<'a> {
    pub name: Name<'a>,
    pub ty: Name<'a>,
    pub init: Option<Expr<'a>>,
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub enum Expr<'a> {
    Assignment(Assignment<'a>),
    Call(Call<'a>),
    If(If<'a>),
    While(While<'a>),
    Block(Block<'a>),
    Let(Let<'a>),
    Case(Case<'a>),
    New(New<'a>),
    BinOp(BinOpExpr<'a>),
    UnOp(UnOpExpr<'a>),
    Name(Name<'a>),
    Int(IntLit<'a>),
    String(StringLit<'a>),
    Bool(BoolLit<'a>),
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct Assignment<'a> {
    pub name: Name<'a>,
    pub expr: Box<Expr<'a>>,
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct Call<'a> {
    pub receiver: Receiver<'a>,
    pub method: Name<'a>,
    pub args: Vec<Expr<'a>>,
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub enum Receiver<'a> {
    SelfType,
    Dynamic(Box<Expr<'a>>),
    Static {
        object: Box<Expr<'a>>,
        ty: Name<'a>,
    },
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct If<'a> {
    pub condition: Box<Expr<'a>>,
    pub consequent: Box<Expr<'a>>,
    pub alternative: Box<Expr<'a>>,
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct While<'a> {
    pub condition: Box<Expr<'a>>,
    pub body: Box<Expr<'a>>,
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct Block<'a> {
    pub body: Vec<Expr<'a>>,
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct Let<'a> {
    pub bindings: Vec<Binding<'a>>,
    pub expr: Box<Expr<'a>>,
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct Case<'a> {
    pub scrutinee: Box<Expr<'a>>,
    pub arms: Vec<CaseArm<'a>>,
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct CaseArm<'a> {
    pub name: Name<'a>,
    pub ty: Name<'a>,
    pub expr: Box<Expr<'a>>,
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct New<'a>(pub Name<'a>);

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct BinOpExpr<'a> {
    pub op: BinOpKind,
    pub lhs: Box<Expr<'a>>,
    pub rhs: Box<Expr<'a>>,
}

define_op_kind!(BinOpKind {
    Add => Plus,
    Subtract => Minus,
    Multiply => Asterisk,
    Divide => Slash,
    LessThan => Less,
    LessEquals => LessEquals,
    Equals => Equals,
});

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct UnOpExpr<'a> {
    pub op: UnOpKind,
    pub expr: Box<Expr<'a>>,
}

define_op_kind!(UnOpKind {
    IsVoid => IsVoid,
    Complement => Tilde,
    Not => Not,
});

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct IntLit<'a>(pub Token<'a>);

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct StringLit<'a>(pub Token<'a>);

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct BoolLit<'a>(pub Token<'a>);
