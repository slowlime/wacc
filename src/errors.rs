use std::error::Error;
use std::fmt::{self, Display};

use crate::position::{HasSpan, Span};

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Level {
    Fatal,
    Error,
    Warn,
    Info,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct DiagnosticMessage {
    pub span: Option<Span>,
    pub message: String,
}

impl DiagnosticMessage {
    pub fn new(message: String) -> Self {
        Self {
            span: None,
            message,
        }
    }

    pub fn new_with_span(span: Span, message: String) -> Self {
        Self::new(message).with_span(span)
    }

    pub fn with_span(self, span: Span) -> Self {
        Self {
            span: Some(span),
            ..self
        }
    }
}

impl Display for DiagnosticMessage {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", &self.message)
    }
}

impl From<String> for DiagnosticMessage {
    fn from(s: String) -> DiagnosticMessage {
        DiagnosticMessage::new(s)
    }
}

#[derive(Debug)]
pub struct Diagnostic {
    pub level: Level,
    pub message: DiagnosticMessage,
    pub source: Option<Box<dyn Error + 'static>>,
}

impl Display for Diagnostic {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", &self.message)
    }
}

impl Error for Diagnostic {
    fn source(&self) -> Option<&(dyn Error + 'static)> {
        self.source.as_deref()
    }
}

#[must_use = "DiagnosticBuilder is useless unless emitted"]
pub struct DiagnosticBuilder<'a, 'e> {
    owner: &'a mut Diagnostics<'e>,
    level: Level,
    message: Option<DiagnosticMessage>,
    source: Option<Box<dyn Error + 'static>>,
}

pub trait SpannedError: Error + HasSpan {}

impl<T: Error + HasSpan> SpannedError for T {}

impl<'a, 'e> DiagnosticBuilder<'a, 'e> {
    fn new(owner: &'a mut Diagnostics<'e>, level: Level) -> Self {
        Self {
            owner,
            level,
            message: None,
            source: None,
        }
    }

    pub fn with_message(mut self, message: impl Into<DiagnosticMessage>) -> Self {
        let message = message.into();

        self.message = Some(message);

        self
    }

    pub fn with_source(self, source: Box<dyn Error + 'static>) -> Self {
        Self {
            source: Some(source),
            ..self
        }
    }

    /// Uses the `error` to fill in the following details of the diagnostic to be emitted:
    /// - the location (unless already set)
    /// - the span (unless already set)
    /// - the message (unless already set)
    /// - the source
    pub fn with_span_and_error(mut self, error: impl SpannedError + 'static) -> Self {
        self.message = self.message.or_else(|| {
            Some(DiagnosticMessage {
                span: Some(error.span().into_owned()),
                message: format!("{}", error),
            })
        });

        self.source = Some(Box::new(error));

        self
    }

    /// Emits the diagnostic.
    ///
    /// Panics if the message is not set.
    pub fn emit(self) {
        let diagnostic = Diagnostic {
            level: self.level,
            message: self.message.expect("message must be set"),
            source: self.source,
        };

        self.owner.emit(diagnostic);
    }
}

fn null_emitter(_: &Diagnostic) {}

pub struct Diagnostics<'e> {
    diagnostics: Vec<Diagnostic>,
    emitter: Box<dyn FnMut(&Diagnostic) + 'e>,
    has_errors: bool,
}

impl<'e> Diagnostics<'e> {
    pub fn new() -> Self {
        Self {
            diagnostics: vec![],
            emitter: Box::new(null_emitter),
            has_errors: false,
        }
    }

    pub fn set_emitter(&mut self, emitter: Box<dyn FnMut(&Diagnostic) + 'e>) {
        self.emitter = emitter;
    }

    pub fn has_errors(&self) -> bool {
        self.has_errors
    }

    pub fn with_level(&mut self, level: Level) -> DiagnosticBuilder<'_, 'e> {
        DiagnosticBuilder::new(self, level)
    }

    pub fn fatal(&mut self) -> DiagnosticBuilder<'_, 'e> {
        self.with_level(Level::Fatal)
    }

    pub fn error(&mut self) -> DiagnosticBuilder<'_, 'e> {
        self.with_level(Level::Error)
    }

    pub fn warn(&mut self) -> DiagnosticBuilder<'_, 'e> {
        self.with_level(Level::Warn)
    }

    pub fn info(&mut self) -> DiagnosticBuilder<'_, 'e> {
        self.with_level(Level::Info)
    }

    fn emit(&mut self, diagnostic: Diagnostic) {
        self.has_errors = self.has_errors || diagnostic.level <= Level::Error;
        (self.emitter)(&diagnostic);
        self.diagnostics.push(diagnostic);
    }

    pub fn into_vec(self) -> Vec<Diagnostic> {
        self.diagnostics
    }
}

impl Default for Diagnostics<'_> {
    fn default() -> Self {
        Self::new()
    }
}
