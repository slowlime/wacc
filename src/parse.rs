mod cursor;
mod lexer;
mod parser;
pub mod token;

pub use cursor::Cursor;
pub use lexer::{Lexer, LexerError};
pub use parser::{Parser, ParserError};
