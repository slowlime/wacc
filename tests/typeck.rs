mod common;

use std::borrow::Cow;
use std::path::Path;

use insta::{assert_debug_snapshot, assert_ron_snapshot};
use paste::paste;
use pretty_assertions::assert_str_eq;
use serde::Serialize;

use wacc::analysis::{self, TypeChecker, TypeckResult};
use wacc::ast;
use wacc::ast::dump::{dump_ast, AstDumpFormat};
use wacc::errors::Diagnostics;
use wacc::parse::{Cursor, Lexer, Parser};
use wacc::position::HasSpan;
use wacc::source::{Source, SourceBuffer};

use self::common::Dump;

fn run_typeck_test_inner(
    code: &'static [u8],
    coolc_expected: &'static [u8],
    filename: &'static str,
    compare_dumps: bool,
) {
    let mut source_buf = SourceBuffer::new();
    let mut source = Source::new(&mut source_buf);
    let source_id = source.load_from_string(
        Path::new("tests")
            .join("typeck")
            .join(filename.to_string() + ".cl"),
        code.to_vec(),
    );
    let source_file = source.get(source_id).unwrap();

    let cursor = Cursor::new(source_file);
    let lexer = Lexer::new(cursor);
    let parser = Parser::new(lexer);
    let ast = parser.parse().unwrap();

    let mut diagnostics = Diagnostics::new();

    let typeck = TypeChecker::new(&mut diagnostics, ast.classes);
    let TypeckResult {
        classes,
        ctx: ty_ctx,
        ..
    } = typeck.resolve();

    if !diagnostics.has_errors() {
        analysis::validate_classes(&source, &ty_ctx, &classes);
        analysis::check_has_main_class(&mut diagnostics, &ty_ctx);
    }

    let failed = diagnostics.has_errors();
    let diagnostics = diagnostics.into_vec();

    #[derive(Serialize, Debug)]
    struct SnapshotData<'a, 'buf> {
        failed: bool,
        classes: &'a [ast::Class<'buf>],
    }

    insta::with_settings!({
        description => String::from_utf8_lossy(code),
    }, {
        assert_ron_snapshot!(filename.to_owned() + "-ast", SnapshotData {
            failed,
            classes: &classes,
        });

        assert_debug_snapshot!(filename.to_owned() + "-diagnostics", &diagnostics);
    });

    let span = classes
        .iter()
        .map(HasSpan::span)
        .map(Cow::into_owned)
        .reduce(|hull, span| hull.convex_hull(&span))
        .unwrap();
    let program = ast::Program { classes, span };

    if compare_dumps {
        let coolc_actual = Dump::from(match failed {
            false => {
                let mut buf = vec![];
                dump_ast(
                    &source,
                    AstDumpFormat::Coolc {
                        ignore_types: false,
                    },
                    program,
                    &mut buf,
                )
                .unwrap();

                buf
            }

            true => b"ERR typeck failed\n".to_vec(),
        });

        let coolc_expected = Dump::from(coolc_expected);
        assert_str_eq!(coolc_actual, coolc_expected);
    }
}

macro_rules! run_typeck_test {
    ($filename:expr) => ({
        run_typeck_test!($filename, true);
    });

    ($filename:expr, $( $args:tt )+) => ({
        let code = include_bytes!(concat!("typeck/", $filename, ".cl"));
        let coolc_expected = include_bytes!(concat!("typeck/", $filename, ".coolc-typeck"));
        run_typeck_test_inner(code, coolc_expected, $filename, $( $args )+);
    });
}

macro_rules! run_typeck_tests {
    {
        $(
            $filename:ident $(
                ($( $args:tt )+)
            )?
        ),+
        $(,)?
    } => {
        $(
            paste! {
                #[test]
                fn [< test_ $filename >]() {
                    run_typeck_test!(stringify!($filename) $(, $( $args )+ )?);
                }
            }
        )+
    };
}

run_typeck_tests! {
    fail_no_main_class,
    fail_no_main_method,
    fail_main_method_wrong_signature,
    fail_main_method_inherited,
    fail_invalid_override_signature,
    fail_covariant_return_tys,
    fail_field_override,
    fail_inherit_non_existent,
    fail_double_class,
    fail_self_field,
    fail_cyclic_inheritance,
    pass_static_dispatch_self,
    pass_override,
    pass_coexisting_field_method_names,
    pass_new,
    pass_self_derived,
    pass_lexical_scope,

    // wacc does not conform to the reference implementation in this aspect:
    // it does not allow to refer to uninitialized fields, making it a hard error
    fail_field_init_referring_to_uninit(false),
}
