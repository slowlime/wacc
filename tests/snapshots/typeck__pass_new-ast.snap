---
source: tests/typeck.rs
description: "class Main {\n    main(): Object {\n        let object : Object <- new Object,\n            int : Int <- new Int,\n            string : String <- new String,\n            io : IO <- new IO,\n            bool : Bool <- new Bool,\n            main : Main <- new Main,\n            self_ : SELF_TYPE <- new SELF_TYPE in 0\n    };\n};\n"
expression: "SnapshotData { failed, classes: &classes }"
---
SnapshotData(
  failed: false,
  classes: [
    Class(
      name: TyName(Name(Spanned(
        value: "Main",
        span: Span(
          start: Position(
            src: Some(SourceId(1)),
            byte: 6,
            line: 1,
            col: 7,
          ),
          end: Position(
            src: Some(SourceId(1)),
            byte: 9,
            line: 1,
            col: 10,
          ),
        ),
      ))),
      inherits: None,
      features: [
        Method(Method(
          name: Name(Spanned(
            value: "main",
            span: Span(
              start: Position(
                src: Some(SourceId(1)),
                byte: 17,
                line: 2,
                col: 5,
              ),
              end: Position(
                src: Some(SourceId(1)),
                byte: 20,
                line: 2,
                col: 8,
              ),
            ),
          )),
          params: [],
          return_ty: TyName(Name(Spanned(
            value: "Object",
            span: Span(
              start: Position(
                src: Some(SourceId(1)),
                byte: 25,
                line: 2,
                col: 13,
              ),
              end: Position(
                src: Some(SourceId(1)),
                byte: 30,
                line: 2,
                col: 18,
              ),
            ),
          ))),
          body: Let(Let(
            binding: Binding(
              name: Name(Spanned(
                value: "object",
                span: Span(
                  start: Position(
                    src: Some(SourceId(1)),
                    byte: 46,
                    line: 3,
                    col: 13,
                  ),
                  end: Position(
                    src: Some(SourceId(1)),
                    byte: 51,
                    line: 3,
                    col: 18,
                  ),
                ),
              )),
              binding_id: SerKey(
                idx: 2,
                version: 1,
              ),
              ty_name: TyName(Name(Spanned(
                value: "Object",
                span: Span(
                  start: Position(
                    src: Some(SourceId(1)),
                    byte: 55,
                    line: 3,
                    col: 22,
                  ),
                  end: Position(
                    src: Some(SourceId(1)),
                    byte: 60,
                    line: 3,
                    col: 27,
                  ),
                ),
              ))),
              init: Some(New(New(
                ty_name: TyName(Name(Spanned(
                  value: "Object",
                  span: Span(
                    start: Position(
                      src: Some(SourceId(1)),
                      byte: 69,
                      line: 3,
                      col: 36,
                    ),
                    end: Position(
                      src: Some(SourceId(1)),
                      byte: 74,
                      line: 3,
                      col: 41,
                    ),
                  ),
                ))),
                span: Span(
                  start: Position(
                    src: Some(SourceId(1)),
                    byte: 65,
                    line: 3,
                    col: 32,
                  ),
                  end: Position(
                    src: Some(SourceId(1)),
                    byte: 67,
                    line: 3,
                    col: 34,
                  ),
                ),
                ty: Resolved(Builtin(Object)),
              ))),
              span: Span(
                start: Position(
                  src: Some(SourceId(1)),
                  byte: 46,
                  line: 3,
                  col: 13,
                ),
                end: Position(
                  src: Some(SourceId(1)),
                  byte: 51,
                  line: 3,
                  col: 18,
                ),
              ),
              ty: Resolved(Builtin(Object)),
            ),
            expr: Let(Let(
              binding: Binding(
                name: Name(Spanned(
                  value: "int",
                  span: Span(
                    start: Position(
                      src: Some(SourceId(1)),
                      byte: 89,
                      line: 4,
                      col: 13,
                    ),
                    end: Position(
                      src: Some(SourceId(1)),
                      byte: 91,
                      line: 4,
                      col: 15,
                    ),
                  ),
                )),
                binding_id: SerKey(
                  idx: 3,
                  version: 1,
                ),
                ty_name: TyName(Name(Spanned(
                  value: "Int",
                  span: Span(
                    start: Position(
                      src: Some(SourceId(1)),
                      byte: 95,
                      line: 4,
                      col: 19,
                    ),
                    end: Position(
                      src: Some(SourceId(1)),
                      byte: 97,
                      line: 4,
                      col: 21,
                    ),
                  ),
                ))),
                init: Some(New(New(
                  ty_name: TyName(Name(Spanned(
                    value: "Int",
                    span: Span(
                      start: Position(
                        src: Some(SourceId(1)),
                        byte: 106,
                        line: 4,
                        col: 30,
                      ),
                      end: Position(
                        src: Some(SourceId(1)),
                        byte: 108,
                        line: 4,
                        col: 32,
                      ),
                    ),
                  ))),
                  span: Span(
                    start: Position(
                      src: Some(SourceId(1)),
                      byte: 102,
                      line: 4,
                      col: 26,
                    ),
                    end: Position(
                      src: Some(SourceId(1)),
                      byte: 104,
                      line: 4,
                      col: 28,
                    ),
                  ),
                  ty: Resolved(Builtin(Int)),
                ))),
                span: Span(
                  start: Position(
                    src: Some(SourceId(1)),
                    byte: 89,
                    line: 4,
                    col: 13,
                  ),
                  end: Position(
                    src: Some(SourceId(1)),
                    byte: 91,
                    line: 4,
                    col: 15,
                  ),
                ),
                ty: Resolved(Builtin(Int)),
              ),
              expr: Let(Let(
                binding: Binding(
                  name: Name(Spanned(
                    value: "string",
                    span: Span(
                      start: Position(
                        src: Some(SourceId(1)),
                        byte: 123,
                        line: 5,
                        col: 13,
                      ),
                      end: Position(
                        src: Some(SourceId(1)),
                        byte: 128,
                        line: 5,
                        col: 18,
                      ),
                    ),
                  )),
                  binding_id: SerKey(
                    idx: 4,
                    version: 1,
                  ),
                  ty_name: TyName(Name(Spanned(
                    value: "String",
                    span: Span(
                      start: Position(
                        src: Some(SourceId(1)),
                        byte: 132,
                        line: 5,
                        col: 22,
                      ),
                      end: Position(
                        src: Some(SourceId(1)),
                        byte: 137,
                        line: 5,
                        col: 27,
                      ),
                    ),
                  ))),
                  init: Some(New(New(
                    ty_name: TyName(Name(Spanned(
                      value: "String",
                      span: Span(
                        start: Position(
                          src: Some(SourceId(1)),
                          byte: 146,
                          line: 5,
                          col: 36,
                        ),
                        end: Position(
                          src: Some(SourceId(1)),
                          byte: 151,
                          line: 5,
                          col: 41,
                        ),
                      ),
                    ))),
                    span: Span(
                      start: Position(
                        src: Some(SourceId(1)),
                        byte: 142,
                        line: 5,
                        col: 32,
                      ),
                      end: Position(
                        src: Some(SourceId(1)),
                        byte: 144,
                        line: 5,
                        col: 34,
                      ),
                    ),
                    ty: Resolved(Builtin(String)),
                  ))),
                  span: Span(
                    start: Position(
                      src: Some(SourceId(1)),
                      byte: 123,
                      line: 5,
                      col: 13,
                    ),
                    end: Position(
                      src: Some(SourceId(1)),
                      byte: 128,
                      line: 5,
                      col: 18,
                    ),
                  ),
                  ty: Resolved(Builtin(String)),
                ),
                expr: Let(Let(
                  binding: Binding(
                    name: Name(Spanned(
                      value: "io",
                      span: Span(
                        start: Position(
                          src: Some(SourceId(1)),
                          byte: 166,
                          line: 6,
                          col: 13,
                        ),
                        end: Position(
                          src: Some(SourceId(1)),
                          byte: 167,
                          line: 6,
                          col: 14,
                        ),
                      ),
                    )),
                    binding_id: SerKey(
                      idx: 5,
                      version: 1,
                    ),
                    ty_name: TyName(Name(Spanned(
                      value: "IO",
                      span: Span(
                        start: Position(
                          src: Some(SourceId(1)),
                          byte: 171,
                          line: 6,
                          col: 18,
                        ),
                        end: Position(
                          src: Some(SourceId(1)),
                          byte: 172,
                          line: 6,
                          col: 19,
                        ),
                      ),
                    ))),
                    init: Some(New(New(
                      ty_name: TyName(Name(Spanned(
                        value: "IO",
                        span: Span(
                          start: Position(
                            src: Some(SourceId(1)),
                            byte: 181,
                            line: 6,
                            col: 28,
                          ),
                          end: Position(
                            src: Some(SourceId(1)),
                            byte: 182,
                            line: 6,
                            col: 29,
                          ),
                        ),
                      ))),
                      span: Span(
                        start: Position(
                          src: Some(SourceId(1)),
                          byte: 177,
                          line: 6,
                          col: 24,
                        ),
                        end: Position(
                          src: Some(SourceId(1)),
                          byte: 179,
                          line: 6,
                          col: 26,
                        ),
                      ),
                      ty: Resolved(Builtin(IO)),
                    ))),
                    span: Span(
                      start: Position(
                        src: Some(SourceId(1)),
                        byte: 166,
                        line: 6,
                        col: 13,
                      ),
                      end: Position(
                        src: Some(SourceId(1)),
                        byte: 167,
                        line: 6,
                        col: 14,
                      ),
                    ),
                    ty: Resolved(Builtin(IO)),
                  ),
                  expr: Let(Let(
                    binding: Binding(
                      name: Name(Spanned(
                        value: "bool",
                        span: Span(
                          start: Position(
                            src: Some(SourceId(1)),
                            byte: 197,
                            line: 7,
                            col: 13,
                          ),
                          end: Position(
                            src: Some(SourceId(1)),
                            byte: 200,
                            line: 7,
                            col: 16,
                          ),
                        ),
                      )),
                      binding_id: SerKey(
                        idx: 6,
                        version: 1,
                      ),
                      ty_name: TyName(Name(Spanned(
                        value: "Bool",
                        span: Span(
                          start: Position(
                            src: Some(SourceId(1)),
                            byte: 204,
                            line: 7,
                            col: 20,
                          ),
                          end: Position(
                            src: Some(SourceId(1)),
                            byte: 207,
                            line: 7,
                            col: 23,
                          ),
                        ),
                      ))),
                      init: Some(New(New(
                        ty_name: TyName(Name(Spanned(
                          value: "Bool",
                          span: Span(
                            start: Position(
                              src: Some(SourceId(1)),
                              byte: 216,
                              line: 7,
                              col: 32,
                            ),
                            end: Position(
                              src: Some(SourceId(1)),
                              byte: 219,
                              line: 7,
                              col: 35,
                            ),
                          ),
                        ))),
                        span: Span(
                          start: Position(
                            src: Some(SourceId(1)),
                            byte: 212,
                            line: 7,
                            col: 28,
                          ),
                          end: Position(
                            src: Some(SourceId(1)),
                            byte: 214,
                            line: 7,
                            col: 30,
                          ),
                        ),
                        ty: Resolved(Builtin(Bool)),
                      ))),
                      span: Span(
                        start: Position(
                          src: Some(SourceId(1)),
                          byte: 197,
                          line: 7,
                          col: 13,
                        ),
                        end: Position(
                          src: Some(SourceId(1)),
                          byte: 200,
                          line: 7,
                          col: 16,
                        ),
                      ),
                      ty: Resolved(Builtin(Bool)),
                    ),
                    expr: Let(Let(
                      binding: Binding(
                        name: Name(Spanned(
                          value: "main",
                          span: Span(
                            start: Position(
                              src: Some(SourceId(1)),
                              byte: 234,
                              line: 8,
                              col: 13,
                            ),
                            end: Position(
                              src: Some(SourceId(1)),
                              byte: 237,
                              line: 8,
                              col: 16,
                            ),
                          ),
                        )),
                        binding_id: SerKey(
                          idx: 7,
                          version: 1,
                        ),
                        ty_name: TyName(Name(Spanned(
                          value: "Main",
                          span: Span(
                            start: Position(
                              src: Some(SourceId(1)),
                              byte: 241,
                              line: 8,
                              col: 20,
                            ),
                            end: Position(
                              src: Some(SourceId(1)),
                              byte: 244,
                              line: 8,
                              col: 23,
                            ),
                          ),
                        ))),
                        init: Some(New(New(
                          ty_name: TyName(Name(Spanned(
                            value: "Main",
                            span: Span(
                              start: Position(
                                src: Some(SourceId(1)),
                                byte: 253,
                                line: 8,
                                col: 32,
                              ),
                              end: Position(
                                src: Some(SourceId(1)),
                                byte: 256,
                                line: 8,
                                col: 35,
                              ),
                            ),
                          ))),
                          span: Span(
                            start: Position(
                              src: Some(SourceId(1)),
                              byte: 249,
                              line: 8,
                              col: 28,
                            ),
                            end: Position(
                              src: Some(SourceId(1)),
                              byte: 251,
                              line: 8,
                              col: 30,
                            ),
                          ),
                          ty: Resolved(Class("Main")),
                        ))),
                        span: Span(
                          start: Position(
                            src: Some(SourceId(1)),
                            byte: 234,
                            line: 8,
                            col: 13,
                          ),
                          end: Position(
                            src: Some(SourceId(1)),
                            byte: 237,
                            line: 8,
                            col: 16,
                          ),
                        ),
                        ty: Resolved(Class("Main")),
                      ),
                      expr: Let(Let(
                        binding: Binding(
                          name: Name(Spanned(
                            value: "self_",
                            span: Span(
                              start: Position(
                                src: Some(SourceId(1)),
                                byte: 271,
                                line: 9,
                                col: 13,
                              ),
                              end: Position(
                                src: Some(SourceId(1)),
                                byte: 275,
                                line: 9,
                                col: 17,
                              ),
                            ),
                          )),
                          binding_id: SerKey(
                            idx: 8,
                            version: 1,
                          ),
                          ty_name: TyName(Name(Spanned(
                            value: "SELF_TYPE",
                            span: Span(
                              start: Position(
                                src: Some(SourceId(1)),
                                byte: 279,
                                line: 9,
                                col: 21,
                              ),
                              end: Position(
                                src: Some(SourceId(1)),
                                byte: 287,
                                line: 9,
                                col: 29,
                              ),
                            ),
                          ))),
                          init: Some(New(New(
                            ty_name: TyName(Name(Spanned(
                              value: "SELF_TYPE",
                              span: Span(
                                start: Position(
                                  src: Some(SourceId(1)),
                                  byte: 296,
                                  line: 9,
                                  col: 38,
                                ),
                                end: Position(
                                  src: Some(SourceId(1)),
                                  byte: 304,
                                  line: 9,
                                  col: 46,
                                ),
                              ),
                            ))),
                            span: Span(
                              start: Position(
                                src: Some(SourceId(1)),
                                byte: 292,
                                line: 9,
                                col: 34,
                              ),
                              end: Position(
                                src: Some(SourceId(1)),
                                byte: 294,
                                line: 9,
                                col: 36,
                              ),
                            ),
                            ty: Resolved(SelfType(
                              enclosed: Class("Main"),
                            )),
                          ))),
                          span: Span(
                            start: Position(
                              src: Some(SourceId(1)),
                              byte: 271,
                              line: 9,
                              col: 13,
                            ),
                            end: Position(
                              src: Some(SourceId(1)),
                              byte: 275,
                              line: 9,
                              col: 17,
                            ),
                          ),
                          ty: Resolved(SelfType(
                            enclosed: Class("Main"),
                          )),
                        ),
                        expr: Int(IntLit(Spanned(
                          value: 0,
                          span: Span(
                            start: Position(
                              src: Some(SourceId(1)),
                              byte: 309,
                              line: 9,
                              col: 51,
                            ),
                            end: Position(
                              src: Some(SourceId(1)),
                              byte: 309,
                              line: 9,
                              col: 51,
                            ),
                          ),
                        ))),
                        span: Span(
                          start: Position(
                            src: Some(SourceId(1)),
                            byte: 271,
                            line: 9,
                            col: 13,
                          ),
                          end: Position(
                            src: Some(SourceId(1)),
                            byte: 275,
                            line: 9,
                            col: 17,
                          ),
                        ),
                      )),
                      span: Span(
                        start: Position(
                          src: Some(SourceId(1)),
                          byte: 234,
                          line: 8,
                          col: 13,
                        ),
                        end: Position(
                          src: Some(SourceId(1)),
                          byte: 237,
                          line: 8,
                          col: 16,
                        ),
                      ),
                    )),
                    span: Span(
                      start: Position(
                        src: Some(SourceId(1)),
                        byte: 197,
                        line: 7,
                        col: 13,
                      ),
                      end: Position(
                        src: Some(SourceId(1)),
                        byte: 200,
                        line: 7,
                        col: 16,
                      ),
                    ),
                  )),
                  span: Span(
                    start: Position(
                      src: Some(SourceId(1)),
                      byte: 166,
                      line: 6,
                      col: 13,
                    ),
                    end: Position(
                      src: Some(SourceId(1)),
                      byte: 167,
                      line: 6,
                      col: 14,
                    ),
                  ),
                )),
                span: Span(
                  start: Position(
                    src: Some(SourceId(1)),
                    byte: 123,
                    line: 5,
                    col: 13,
                  ),
                  end: Position(
                    src: Some(SourceId(1)),
                    byte: 128,
                    line: 5,
                    col: 18,
                  ),
                ),
              )),
              span: Span(
                start: Position(
                  src: Some(SourceId(1)),
                  byte: 89,
                  line: 4,
                  col: 13,
                ),
                end: Position(
                  src: Some(SourceId(1)),
                  byte: 91,
                  line: 4,
                  col: 15,
                ),
              ),
            )),
            span: Span(
              start: Position(
                src: Some(SourceId(1)),
                byte: 42,
                line: 3,
                col: 9,
              ),
              end: Position(
                src: Some(SourceId(1)),
                byte: 44,
                line: 3,
                col: 11,
              ),
            ),
          )),
          span: Span(
            start: Position(
              src: Some(SourceId(1)),
              byte: 17,
              line: 2,
              col: 5,
            ),
            end: Position(
              src: Some(SourceId(1)),
              byte: 20,
              line: 2,
              col: 8,
            ),
          ),
          ty: Resolved(Function(FunctionTy(
            params: [],
            ret: Builtin(Object),
          ))),
        )),
      ],
      span: Span(
        start: Position(
          src: Some(SourceId(1)),
          byte: 0,
          line: 1,
          col: 1,
        ),
        end: Position(
          src: Some(SourceId(1)),
          byte: 4,
          line: 1,
          col: 5,
        ),
      ),
      ty: Resolved(Class("Main")),
    ),
  ],
)
